### PR TITLE
Distance API added

### LINKED ISSUE #168: Distance API support
This suports the Directions API, but not the Distances (Directions Matrix) API

### PR DESCRIPTION
**Pull request by cammace**
Closes #168

Adds the [Distance API](https://www.mapbox.com/api-documentation/?language=Java#distance) to mapbox-java. Still need to add test and clean up some of them code. Example activity can be found in testapp:

cc: @mapbox/android

### CODE REVIEW

**Comment by cammace** [comment]
Build is failing since I upgraded from `4.2.0-beta.1` to `4.2.0-beta.5` which includes libjava in it. Dex files were interfering. I added 
```groovy
exclude group: 'com.mapbox.mapboxsdk', module: 'mapbox-java-services'
```

**Comment by ivovandongen** [comment]
@cammace The Distance related code seems good. There is a lot in there that doesn't seem to belong to this PR though (images, dependency updates). Would be great to separate these for easier reviews/traceability.

**Comment by cammace** [comment]
Cleaned up PR and moved some changes into https://github.com/mapbox/mapbox-java/pull/223.

File: libandroid/app/src/main/java/com/mapbox/services/android/testapp/distance/DistanceActivity.java [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+  @Override
+  protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_distance);
+
+    Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
+    setSupportActionBar(toolbar);
+    getSupportActionBar().setDisplayHomeAsUpEnabled(true);
+
+    coordinates = new ArrayList<>();
```
Comment (ivovandongen) [inline libandroid/app/src/main/java/com/mapbox/services/android/testapp/distance/DistanceActivity.java]: Initialisation can go into the field declaration as it's only done once anyway. That way you can make them final and there's no chance of null pointer exceptions (when refactoring)

File: libandroid/app/src/main/java/com/mapbox/services/android/testapp/distance/DistanceActivity.java [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+    }
+
+    addMarkers();
+    updateLayers();
+  }
+
+  @Override
+  public void onMapClick(@NonNull LatLng point) {
+
+    features = new ArrayList<>();
```
Comment (ivovandongen) [inline libandroid/app/src/main/java/com/mapbox/services/android/testapp/distance/DistanceActivity.java]: replace with `clear` and initialise in declaration

File: libandroid/app/src/main/java/com/mapbox/services/android/testapp/distance/DistanceActivity.java [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+    for (int i = 0; i < restaurants.size(); i++) {
+      features.addAll(mapboxMap.queryRenderedFeatures(
+        pixel,
+        restaurants.get(i).getStringProperty(KEY_NAME) + "-layer"
+      ));
+    }
+
+    features.addAll(mapboxMap.queryRenderedFeatures(pixel, "circle-layer"));
+
+    if (features.size() <= 0) {
```
Comment (ivovandongen) [inline libandroid/app/src/main/java/com/mapbox/services/android/testapp/distance/DistanceActivity.java]: Might want to do the opposite here; `size() > 0` -> `callDistanceApi()` and have just a single exit point in the method

File: libandroid/app/src/main/java/com/mapbox/services/android/testapp/distance/DistanceActivity.java [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+              if (restaurants.get(i).getProperty(KEY_NAME).getAsString().equals(
+                features.get(0).getStringProperty(KEY_NAME))) {
+                feature = i;
+              }
+            }
+
+            for (int i = 0; i < restaurants.size(); i++) {
+              // Get the json object and replace the duration property with the updated time. We convert the API's
+              // second value to minutes.
+              JsonObject jsonObject = restaurants.get(i).getProperties().getAsJsonObject();
```
Comment (ivovandongen) [inline libandroid/app/src/main/java/com/mapbox/services/android/testapp/distance/DistanceActivity.java]: Can never be null?

File: libandroid/app/src/main/java/com/mapbox/services/android/testapp/distance/DistanceActivity.java [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+        PropertyFactory.circleRadius(10f)
+      );
+
+      mapboxMap.addLayer(circleLayer);
+
+      SymbolLayer nameLayer = new SymbolLayer(
+        restaurants.get(i).getStringProperty("name") + "-layer-name",
+        restaurants.get(i).getStringProperty("name") + "-source"
+      ).withProperties(
+        PropertyFactory.textSize(14f),
```
Comment (ivovandongen) [inline libandroid/app/src/main/java/com/mapbox/services/android/testapp/distance/DistanceActivity.java]: Can be shortened with static import

File: libjava/lib/src/main/java/com/mapbox/services/distance/v1/gson/DistanceGeometryDeserializer.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+   * @param typeOfT The type of the Object to deserialize to.
+   * @param context Context for deserialization.
+   * @return Deserialized Geometry.
+   * @throws JsonParseException This exception is raised if there is a serious issue that occurs
+   *                            during parsing of a Json string.
+   * @since 2.0.0
+   */
+  @Override
+  public Geometry deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
+    throws JsonParseException {
```
Comment (ivovandongen) [inline libjava/lib/src/main/java/com/mapbox/services/distance/v1/gson/DistanceGeometryDeserializer.java]: Indentation looks a bit off here. Or is that the new code style?

Comment (cammace) [inline libjava/lib/src/main/java/com/mapbox/services/distance/v1/gson/DistanceGeometryDeserializer.java]: it's the new style, line 31 is over the max line length so I broke it up by adding `throws` on the line below.

**Pull request merged**